### ISSUE TITLE
Inline JS of zimit ZIMs is not working properly

### ISSUE THREAD

**Issue by benoit74** [kiwix/kiwix-desktop#1281]
https://github.com/kiwix/kiwix-desktop/issues/1276 has revealed that on both Windows and Linux, all tests linked to inline JS are not working properly.

These are tests 01* of "Javascript" page of the ZIM available at https://dev.library.kiwix.org/#lang=eng&q=zimit+test+website

Do not mind about tests 08* and 09*, they are known to not be reliable yet due to other issues not yet solved in zimit/warc2zim.

These tests are known to work ok on all readers (kiwix-desktop is probably the last one not yet tested), at least I'm 100% sure they do work on kiwix-android, kiwix-apple, kiwix-serve, kiwix-js

These tests are very specific to zimit/warc2zim. They rely on proper operation of wombat, the JS library used by zimit/warc2zim ZIMs to replace dynamically / on-the-fly URLs loaded by JS code. It is hence "normal" that the URL inside JS code is "online", it will be rewritten by wombat to the proper in-ZIM URL.

**Comment by kelson42** [comment]
OK, if this is a blocker to get Zimit video working, I guess this is the new TOP priority issue.

**Comment by benoit74** [comment]
> OK, if this is a blocker to get Zimit video working, I guess this is the new TOP priority issue.

It looks it is a very distinct issue

**Comment by kelson42** [comment]
@benoit74 OK, so far we have done our best to remove all inline JS because this is not supported in Kiwix JS (and there is - looks like - no way to fix this). @Jaifroid could confirm. We had an effort specifically about that 2 years ago. This is actually still a constraint applying on all our custom scrapers if we want to continue to have things working on Kiwix JS. This is very important.

This is the reason why this limitation of Kiwix Desktop has not been considered properly. But if this impacts ZIMit, where we have less control about the page content, then we will have to find a fix here with high priority.

**Comment by Jaifroid** [comment]
> [#1276](https://github.com/kiwix/kiwix-desktop/issues/1276) has revealed that on both Windows and Linux, all tests linked to inline JS are not working properly.
> 
> These tests are known to work ok on all readers (kiwix-desktop is probably the last one not yet tested), at least I'm 100% sure they do work on kiwix-android, kiwix-apple, kiwix-serve, kiwix-js
> 
> These tests are very specific to zimit/warc2zim. They rely on proper operation of wombat, the JS library used by zimit/warc2zim ZIMs to replace dynamically / on-the-fly URLs loaded by JS code. It is hence "normal" that the URL inside JS code is "online", it will be rewritten by wombat to the proper in-ZIM URL.

@benoit74 It's perfectly fine to use the PWA (pwa.kiwix.org or browser-extension.kiwis.org, either is fine) to run those tests, as they are tests of Zimit, which by definition requires a lot of inline JS which is fully supported by PWAs. So no problem there on my side.

**Comment by veloman-yunkan** [comment]
While opening in `kiwix-desktop` the Javascript page of the "Zimit test website" ZIM file the following output is generated in the terminal:

```
js: You are using the fallback version (version 2.0.0-dev9) of wombatSetup.js, this is not expected outside dev
js: 'window.webkitStorageInfo' is deprecated. Please use 'navigator.webkitTemporaryStorage' or 'navigator.webkitPersistentStorage' instead.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/website.test.openzim.org/javascript/content.txt. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/w%C3%A9bsite.test.openzim.org/javascript/content.txt. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/website.test.openzim.org/javascript/cont%C3%A9nt.txt. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/w%C3%A9bsite.test.openzim.org/javascript/cont%C3%A9nt.txt. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/website.test.openzim.org/javascript/cont%F0%9F%8E%81nt.txt. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/w%C3%A9bsite.test.openzim.org/javascript/cont%F0%9F%8E%81nt.txt. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/website.test.openzim.org/javascript/cont%C3%A9nt.txt. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/w%C3%A9bsite.test.openzim.org/javascript/cont%C3%A9nt.txt. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/w%C3%A9bsite.test.openzim.org/javascript/content.txt. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/website.test.openzim.org/javascript/content.txt%3Fquery%3Dvalue. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/website.test.openzim.org/javascript/content.txt%3Fquery%3Dvalu%3De. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/website.test.openzim.org/javascript/content.txt%3Fquery%3Dvalu%25e. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/website.test.openzim.org/javascript/cont%3Fnt.txt. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/website.test.openzim.org/javascript/cont%3Fnt.txt%3Fquery%3Dvalue. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/website.test.openzim.org/javascript/cont%F0%9F%8E%81nt.txt. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/w%C3%A9bsite.test.openzim.org/javascript/cont%F0%9F%8E%81nt.txt. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/website.test.openzim.org/javascript/cont%F0%9F%8E%81nt.txt. URL scheme "zim" is not supported.
js: Fetch API cannot load zim://7aaf6cc6-e68e-fbba-af18-da96f0004aa3.zim/website.test.openzim.org/javascript/cont!nt.txt. URL scheme "zim" is not supported.
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught (in promise) TypeError: Failed to fetch
js: Uncaught TypeError: Failed to resolve module specifier "resources". Relative references must start with either "/", "./", or "../".
js: Uncaught TypeError: Failed to resolve module specifier "javascript/resources". Relative references must start with either "/", "./", or "../".
```

Now going to debug why the URL scheme "zim://" isn't available to the Fetch API.

**Comment by kelson42** [comment]
> Now going to debug why the URL scheme "zim://" isn't available to the Fetch API.

See also https://github.com/kiwix/kiwix-desktop/issues/917

**Comment by veloman-yunkan** [comment]
Looks like the only fix to this issue is to use Qt 6.6 or newer - see https://stackoverflow.com/questions/64892161/qtwebengine-fetch-api-fails-with-custom-scheme

**Comment by Jaifroid** [comment]
AFAIK the Fetch API requires a secure context (HTTPS) when used with modern browsers, with a few specific exceptions (requests to localhost and 127.0.0.1, file:// URLs, dev environments on port 80). Fetch is closely tied to the Service Worker spec, and secure environment prevents man-in-the-middle attacks, etc.

To support Fetch fully, you might need to switch to having an internal server. This is very similar to what I had to do with the Electron KJS app, as there is no way to get some zimit-based content working when using from chrome:// and moz-extension:// URL schemata.

For me there were only two solutions:

* Serve the entire app from a secure server (https://pwa.kiwix.org, or https://browser-extension.kiwix.org) - solution used in Browser Extension and for standalone PWA
* Serve from http://localhost which is considered secure for dev purposes - solution used in the Electron apps.

**Comment by veloman-yunkan** [comment]
> Would you be able to implement the fix based on qt6+?
> 
> FYI, once 2.4 release, the first thing I would like to do would be to move to QT6 as default (only?) version and library for our binaries.

@kelson42 Note that switching to qt6 is not enough. For example, on Ubuntu 22.04 jammy (our next likely default build platform after Ubuntu 20.04 focal's standard security maintenance five year period ends) the version of Qt 6 is 6.2.4 (the needed fix first appears in Qt 6.6).

**Comment by kelson42** [comment]
> > Would you be able to implement the fix based on qt6+?
> > FYI, once 2.4 release, the first thing I would like to do would be to move to QT6 as default (only?) version and library for our binaries.
> 
> [@kelson42](https://github.com/kelson42) Note that switching to qt6 is not enough. For example, on Ubuntu 22.04 jammy (our next likely default build platform after Ubuntu 20.04 focal's standard security maintenance five year period ends) the version of Qt 6 is 6.2.4 (the needed fix first appears in Qt 6.6).

I understand that we won't have a 100% fix even by fixing right now, but (1) we have to move to Qt6 to allow to fix the bug (2) This is important bug which has to be fixed in prio (3) We have to live with what we can't change - at least we are ready.

**Comment by kelson42** [comment]
Blocked by https://github.com/kiwix/kiwix-build/issues/594

**Comment by benoit74** [comment]
I confirm that the Fetch API is not working properly in kiwix-desktop, I fixed freecodecamp ZIMs by removing it, see #1378

Now that move to Qt6 has been done, looks pretty important to move this forward IMHO.

Ubuntu Focal is not out of security maintenance, so our target seems to be Jammy. However, are we still blocked by the fact that Jammy still targets Qt 6.2 and we need Qt 6.6? What will happen if we add the "magic" flag to make Fetch API work but still run on 6.2? The flag will not exist so we will have an error?

**Comment by kelson42** [comment]
@jasontitus @veloman-yunkan Duplicate of #1476?

**Comment by benoit74** [comment]
Note that this can be easily confirmed by using a recent QT6.6+ build of kiwix-desktop with my original comment https://github.com/kiwix/kiwix-desktop/issues/1281#issue-2731049018

I don't have such a build at hand unfortunately (and can't build one easily).

**Comment by veloman-yunkan** [comment]
> [@jasontitus](https://github.com/jasontitus) [@veloman-yunkan](https://github.com/veloman-yunkan) Duplicate of [#1476](https://github.com/kiwix/kiwix-desktop/issues/1476)?

@kelson42 I'd rather say that #1476 (the newer ticket) is a duplicate of this one.
